datafusion/expr/src/expr.rs

```diff
@@ -1692,6 +1692,24 @@ fn create_names(exprs: &[Expr]) -> Result<String> {
         .join(", "))
 }
 
+/// Whether the given expression is volatile, i.e. whether it can return different results
+/// when evaluated multiple times with the same input.
+#[allow(clippy::match_like_matches_macro)]
+pub fn is_volatile(expr: &Expr) -> bool {
+    match expr {
+        Expr::ScalarFunction(func) => match func.func_def {
+            ScalarFunctionDefinition::BuiltIn(func) => match func {
+                BuiltinScalarFunction::Random => true,
+                _ => false,
+            },
+            // TODO: Add volatile flag to UDFs
+            ScalarFunctionDefinition::UDF(_) => false,
```

Comment:
Ideally, UDFs could have volatile variants. I defer to work on it later and focus this PR for the bug fix only.

Reply:
I think they can have volatile set as part of their signature: https://docs.rs/datafusion/latest/datafusion/logical_expr/struct.ScalarUDF.html#structfield.signature
 (and a field on Signature is `volatile`)